### DISCUSSION TITLE
Retrieve invoice data??

### DISCUSSION THREAD

**Discussion by Ojay** [swellstores/community#82] [Ideas]
Hi there, I asked this question [back in the old forum](https://community.swell.is/discussion/611/retrieve-invoice-data-with-swell-js) but the answer doesn't seem to be returning any payments/invoices.

Essentially, I'm just looking to pull the invoices/payments associated with a Subscription. The referenced code (also pasted below) returns an empty `payments.results` array. Any update on this or..?

```

    await swell.account.login('email', 'password');
    let total = 0;

    const subscriptionResults = await swell.subscriptions.get({
      id: '62a0af3146ff521a438ad62d',
      expand: ['orders:0'],
    })

    const subscription = subscriptionResults?.results?.length ? subscriptionResults.results[0] : null;
    if (subscription) {
      if (subscription?.orders?.results) {
        // get each order with its payments
        for (const order of subscription.orders.results) {
          const orderWithPaymentsResults = await swell.get('/orders', {
            id: order.id,
            expand: ['payments:0'],
          });

          const orderWithPayments = orderWithPaymentsResults?.results?.length ? orderWithPaymentsResults.results[0] : null;
          if (orderWithPayments?.payments?.results) {
            for (const payment of orderWithPayments.payments.results) {
              // handle this payment
              console.log(payment);
              if (payment.status === 'success') {
                total = total + payment.amount;
              }
            }
          }
        }
      }
    }
```

**Comment by kyle-swell** [comment]
Hi @Ojay. Does this subscription you are viewing have orders associated to it? It may instead have invoices instead of orders. If it has invoices you can retrieve the payments from each of the invoices instead in the same way you are trying to get them from the orders.

**Reply by Ojay** [reply]
Thanks for the prompt response @kyle-swell - appreciate it. The goal of that code was to get the invoices, but was a bit of a hack as the `invoices` api wasn't yet available in `swell-js` apparently. Updating the code as you suggest returns an api permissions error...

`error Error: Permissions for invoices must be defined in api.json`

```
const subscriptionsList = await swell.subscriptions.list({
        expand: ['orders:0'],
      });

      const acc = await swell.account.get();

      const subscription = subscriptionsList?.results?.length
        ? subscriptionsList.results[0]
        : null;
      if (subscription) {
        if (subscription?.orders?.results) {
          // get each order with its payments
          for (const order of subscription.orders.results) {
            const orderWithPaymentsResults = await swell.get('/invoices/{id}', {
              id: acc.id,
            });
            console.log({orderWithPaymentsResults});
            const orderWithPayments = orderWithPaymentsResults?.results?.length
              ? orderWithPaymentsResults.results[0]
              : null;
            if (orderWithPayments?.payments?.results) {
              for (const payment of orderWithPayments.payments.results) {
                // handle this payment
                
                if (payment.status === 'success') {
                  total += payment.amount;
                }
              }
            }
          }
        }
      }
```

**Reply by ericingram** [reply]
We should add default permissions for invoices @kyle-swell

**Reply by kyle-swell** [reply]
Sorry about that @Ojay . Can you try expanding payments directly from the subscriptions GET request.

`
const subscriptionsList = await swell.subscriptions.list({
        expand: ['payments'],
      });
`
If you have permission to do that you should be able to iterate over the payments to get your total payment amount for a subscription. Let me know if that doesn't work for you.

**Reply by Ojay** [reply]
Thanks again @kyle-swell for the efforts but that returns the following error...

`Error: You can only expand fields configured in api.json (payments)`

**Reply by Ojay** [reply]
Any further thoughts on this @kyle-swell or @ericingram? I've got a bit of time before we launch but will need some kind of a solution soon... I'll keep looking into it but it very much seems its more of a permissions issue, which I obviously have no control over.

Thanks so much for your time and help so far.

**Reply by osseonews** [reply]
I might be wrong, but I don't think you can get invoices from the API. I ran into the same issue 10 months ago and requested this. Ended up having to just custom code our own invoices. We get the order information, save it back to our own database, and then use another API to generate the invoices.

**Reply by Ojay** [reply]
Thanks for getting back to me @osseonews - sorry you couldn't find an appropriate solution.
Its a weird one because it seems grabbing Invoice data in `swell-node` is very straightforward. Considering @ericingram's comment about default permissions, it seems like the functionality is there but it's just not been switched on. Is that right or..?

**Reply by kyle-swell** [reply]
@Ojay Apologies for the late response. We looked further into this and that is correct, the permissions defined in our storefront api currently do not allow access to subscription invoices or payments. We are looking into adding access to those models now.

**Reply by Ojay** [reply]
Okay great, thanks for letting me know @kyle-swell. I'm okay for now but am hoping to launch in March and so fingers crossed its sorted for then! :-)

Thanks for all your help and especially for making such an epic product.

**Comment by swellmike** [comment]
We've added customer invoices and subscription payments to the Storefront API.  In addition, [swell-js](https://www.npmjs.com/package/swell-js) v3.19.8  should now have support for fetching customer invoices.

Hope that helps!
